File: cairnq-node/src/ids.ts
import { randomBytes } from "node:crypto";

// ULID-style id. Must match the Python SDK byte-for-byte in format (PROTOCOL.md):
// <prefix>_ + 26-char Crockford base32 of (48-bit ms timestamp << 80 | 80-bit random).
const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

export function newUlid(tsMs: number = Date.now()): string {
  let value = (BigInt(tsMs) << 80n) | BigInt("0x" + randomBytes(10).toString("hex"));
  const chars: string[] = [];
  for (let i = 0; i < 26; i++) {
    chars.push(CROCKFORD[Number(value & 31n)]);
    value >>= 5n;
  }
  return chars.reverse().join("");
}

export function newId(prefix = "task"): string {
  return `${prefix}_${newUlid()}`;
}

export function nowMs(): number {
  return Date.now();
}

File: cairnq-py/cairnq/_ids.py
"""ULID-style id generation. Format pinned by PROTOCOL.md and asserted by
conformance: `<prefix>_` + 26-char Crockford base32 (48-bit ms timestamp +
80-bit randomness), lexicographically time-sortable. Must match the TS SDK."""

from __future__ import annotations

import os
import time

_CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


def new_ulid(ts_ms: int | None = None) -> str:
    if ts_ms is None:
        ts_ms = int(time.time() * 1000)
    value = (ts_ms << 80) | int.from_bytes(os.urandom(10), "big")
    chars = []
    for _ in range(26):
        chars.append(_CROCKFORD[value & 0x1F])
        value >>= 5
    return "".join(reversed(chars))


def new_id(prefix: str = "task") -> str:
    return f"{prefix}_{new_ulid()}"


def now_ms() -> int:
    return int(time.time() * 1000)
